### PR TITLE
Force driver to use Vert.x event loop 

### LINKED ISSUE #3: Integrate Vert.x and driver's event loop

### PR DESCRIPTION
**Pull request by Sammers21**
Fixes #3

### CODE REVIEW

**Comment by Sammers21** [comment]
@vietj , @tsegismont , can you review?

**Comment by Sammers21** [comment]
You mean in which thread we are when a callback is called?

**Comment by Sammers21** [comment]
This is how driver's bootstrap is configured: https://github.com/datastax/java-driver/blob/3.x/driver-core/src/main/java/com/datastax/driver/core/Connection.java#L1032

**Comment by vietj** [comment]
I was actually referring to the code called back from Cassandra like:

```
vertxExecuteFuture.setHandler(executionResult -> {
   // vertx code
});
```

in the final callback to the user code:

- do we check that the event loop thread is the same
- the vertx context is the same as well

do you use `ContextInternal#executeFromIO` before calling user code ?

**Comment by vietj** [comment]
cf this : https://github.com/vietj/vertx-materials/blob/master/src/main/asciidoc/output/Integrating_Netty.adoc

**Comment by Sammers21** [comment]
Ah, I got it. Sure, we need to call `ContextInternal#executeFromIO` before invoking application code

**Comment by Sammers21** [comment]
I also think that driver should use acceptor event loop because we can't specify one for networking and one for calling handlers

**Comment by Sammers21** [comment]
@vietj , can you review, please?

**Comment by Sammers21** [comment]
@vietj , ping

**Comment by vietj** [comment]
I think we should have tests with callbacks that:

1/ check the context is the right one
2/ throwing an exception in the handler is actually reported on the `Context#exceptionHandler`

**Comment by Sammers21** [comment]
@vietj , done. Can you review one more time, please?

File: src/main/java/io/vertx/cassandra/impl/CassandraClientImpl.java [hunk 1]
```diff
@@ -212,17 +225,18 @@ public CassandraClient prepare(String query, Handler<AsyncResult<PreparedStateme
...
       Future<com.datastax.driver.core.PreparedStatement> vertxExecuteFuture = Util.toVertxFuture(preparedFuture, vertx);
-      vertxExecuteFuture.setHandler(executionResult -> {
-        if (executionResult.succeeded()) {
-          if (resultHandler != null) {
-            resultHandler.handle(Future.succeededFuture(executionResult.result()));
-          }
-        } else {
-          if (resultHandler != null) {
-            resultHandler.handle(Future.failedFuture(executionResult.cause()));
+      vertxExecuteFuture.setHandler(executionResult ->
```
Comment (vietj) [inline src/main/java/io/vertx/cassandra/impl/CassandraClientImpl.java]: you can simplify to :

```
    vertxExecuteFuture.setHandler(executionResult -> {
      if (resultHandler != null) {
        ctx.executeFromIO(executionResult, resultHandler);
      }
    });
```

**Comment by vietj** [comment]
there are simplifications you can do in callback, there is a method on `ContextInternal` that allows to run `executeFromIO` using an `Handler<T>` that takes a `T` as parameter you can use. Also re-creating a `Future` from a `Future` is not necessary. Take a look.

File: src/main/java/io/vertx/cassandra/impl/CassandraClientImpl.java (hunk 1 above)
Comment (Sammers21) [reply src/main/java/io/vertx/cassandra/impl/CassandraClientImpl.java]: Done

**Comment by Sammers21** [comment]
@vietj, can you review once again?

**Comment by Sammers21** [comment]
@tsegismont , can you review, please?

File: src/test/java/io/vertx/cassandra/ConnectionTest.java [hunk 2]
```diff
@@ -50,10 +50,12 @@ private void connectAndDisconnect(TestContext context, CassandraClientOptions op
     );
     Async async = context.async(2);
     cassandraClient.connect("system", connectEvent -> {
+      checkContext(context);
```
Comment (tsegismont) [inline src/test/java/io/vertx/cassandra/ConnectionTest.java]: I don't believe it is necessary to do this in all the tests in the test suite. Instead, we could have a test dedicated to context checking (with different use cases for shared/non shared client)

**Comment by Sammers21** [comment]
@tsegismont, can you review, please?

File: src/test/java/io/vertx/cassandra/ConnectionTest.java (hunk 2 above)
Comment (Sammers21) [reply src/test/java/io/vertx/cassandra/ConnectionTest.java]: > I don't believe it is necessary to do this in all the tests in the test suite. Instead, we could have a test dedicated to context checking

Agree. 

> with different use cases for shared/non shared client

Why do we need to have context checks for both of them?

**Comment by Sammers21** [comment]
It seems that there is something wrong with this approach. We are getting IllegalStateEception:

```
java.lang.IllegalStateException: Uh oh! Event loop context executing with wrong thread! Expected null got Thread[main,5,main]
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:318)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:188)
	at io.vertx.core.impl.EventLoopContext.executeFromIO(EventLoopContext.java:23)
	at io.vertx.cassandra.impl.CassandraRowStreamImpl.tryToTriggerEndOfTheStream(CassandraRowStreamImpl.java:134)
	at io.vertx.cassandra.impl.CassandraRowStreamImpl.endHandler(CassandraRowStreamImpl.java:82)
	at io.vertx.cassandra.StreamingTest.lambda$emptyStream$11(StreamingTest.java:95)
```


```
java.lang.IllegalStateException: Uh oh! Event loop context executing with wrong thread! Expected null got Thread[main,5,main]
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:318)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:188)
	at io.vertx.core.impl.EventLoopContext.executeFromIO(EventLoopContext.java:23)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:179)
	at io.vertx.core.impl.EventLoopContext.executeFromIO(EventLoopContext.java:23)
	at io.vertx.cassandra.impl.CassandraClientImpl.lambda$null$15(CassandraClientImpl.java:253)
	at io.vertx.core.impl.FutureImpl.setHandler(FutureImpl.java:79)
	at io.vertx.cassandra.impl.CassandraClientImpl.lambda$disconnect$16(CassandraClientImpl.java:253)
	at io.vertx.cassandra.impl.CassandraClientImpl.executeWithSession(CassandraClientImpl.java:262)
	at io.vertx.cassandra.impl.CassandraClientImpl.disconnect(CassandraClientImpl.java:251)
	at io.vertx.cassandra.ExecutionTest.lambda$simpleReleaseVersionSelect$8(ExecutionTest.java:124)
```

https://travis-ci.org/vert-x3/vertx-cassandra-client/jobs/437563597

**Comment by Sammers21** [comment]
@vietj , @tsegismont do you have any ideas on what is wrong?

**Comment by Sammers21** [comment]
I am also not able to reproduce it on my on my laptop (tried 10 times). Look wired.

**Comment by tsegismont** [comment]
I can reproduce on my box with a vertx-unit repeat rule. Working on the fix.

**Comment by tsegismont** [comment]
@Sammers21 the issue is here: https://github.com/vert-x3/vertx-cassandra-client/blob/7f7fb20e45eecdef8d4b12702cc15adf34914831/src/main/java/io/vertx/cassandra/impl/CassandraClientImpl.java#L160-L164

When `sessionFuture.setHandler` is invoked, it is possible that the future is already completed. Then the calling thread will invoke the handler. In our case, the calling thread is the Java main thread.

The issue is not specific to this PR, it is simply brought forward because `executeFromIO` does thread checks.

**Comment by tsegismont** [comment]
@Sammers21 I overlooked key changes the first time I reviewed this PR. For this to work we must be sure that the Datastax driver callbacks will always be invoked on the connection event loop. I don't believe this is true.

**Comment by vietj** [comment]
you can use EventExecutor#inEventLoop() to check wether yo uare in event loop or not.

Julien

**Comment by tsegismont** [comment]
@vietj right. But my point was that chances are low that the event loop running the driver connection is the same as the event loop running the calling context.

In the end, to make this right, we'll add a lot of complexity for a small benefit.

@Sammers21 @vietj I just submitted #9 and I would rather close this one

**Comment by Sammers21** [comment]
@tsegismont, the #9 PR is fine to me. But, can we make the problem more explicit? A Vert.x instance maintain the acceptor event loop, which is a single-threaded event loop. The event loop should be used for network operations and never be used for executing user code. Right? 

Why can't we use the acceptor event loop for network operations with Cassandra and delegate "calling handlers" to a context related event loop, so user code(inside one verticle) will always be executed in the same context.

**Comment by tsegismont** [comment]
> A Vert.x instance maintain the acceptor event loop, which is a single-threaded event loop.

The `acceptorEventLoopGroup` is a group of a single event loop:

https://github.com/eclipse-vertx/vert.x/blob/9cafc3673424722c463d3e9d3754821f1bb86484/src/main/java/io/vertx/core/impl/VertxImpl.java#L153-L155

Its purpose is to avoid having connections rejected because all the event loops used for verticles are too busy. It should not be reused in the Cassandra client to create event loop for driver connections.

https://github.com/vert-x3/vertx-cassandra-client/blob/7f7fb20e45eecdef8d4b12702cc15adf34914831/src/main/java/io/vertx/cassandra/impl/CassandraClientImpl.java#L280-L283

The code above is wrong. If anything, we should use `VertxInternal#getEventLoopGroup`

> The event loop should be used for network operations and never be used for executing user code. Right?

Not sure what you mean.

> Why can't we use the acceptor event loop for network operations with Cassandra and delegate "calling handlers" to a context related event loop, so user code(inside one verticle) will always be executed in the same context.

I believe I answered the question related to `acceptor`. As for using Vert.x event loops, it sounded to me like a good idea in the beginning, but taking a closer look I can only see problems but not much benefits.

The problems are:
- drivers connections are handled by a single event loop
- there is no guarantee that a verticle using our client will always be assigned the same event loop

Therefore, the benefits (a bit less threads) are not going to outweigh the added complexity.

**Comment by tsegismont** [comment]
@Sammers21 I'll go ahead and merge #9 

We can either close #3, or keep it open and try to demonstrate the benefits with a benchmark?

**Comment by Sammers21** [comment]
@tsegismont , I we can close this PR, and keep the #3 issue open.

**Comment by Sammers21** [comment]
> try to demonstrate the benefits with a benchmark

@tsegismont, yes I think we need to measure performance benefits

**Pull request closed without merging**